homeassistant/auth/mfa_modules/totp.py

```diff
@@ -0,0 +1,142 @@
+
+        self._users = data.get(STORAGE_USERS, [])
+
+    async def _async_save(self) -> None:
+        """Save data."""
+        await self._user_store.async_save({STORAGE_USERS: self._users})
+
+    def _add_ota_secret(self, user_id: str) -> str:
+        """Create a ota_secret for user."""
+        import pyotp
+
+        ota_secret = pyotp.random_base32()  # type: str
+
+        for user in self._users:  # type: ignore
+            if user[STORAGE_USER_ID] == user_id:
+                # found user, override ota secret
+                user[STORAGE_OTA_SECRET] = ota_secret
+                return ota_secret
+
+        self._users.append({  # type: ignore
+            STORAGE_USER_ID: user_id,
+            STORAGE_OTA_SECRET: ota_secret
+        })
+        return ota_secret
+
+    async def async_setup_user(self, user_id: str, setup_data: Any) -> str:
+        """Setup auth module for user."""
+        if self._users is None:
+            await self._async_load()
+
+        result = await self.hass.async_add_executor_job(
+            self._add_ota_secret, user_id)
+
+        await self._async_save()
+        return result
+
+    async def async_depose_user(self, user_id: str) -> None:
+        """Depose auth module for user."""
+        if self._users is None:
+            await self._async_load()
+
+        found = None
+        for user in self._users:  # type: ignore
+            if user[STORAGE_USER_ID] == user_id:
+                found = user
+                break
+        if found:
+            self._users.remove(found)  # type: ignore
+        await self._async_save()
+
+    async def async_is_user_setup(self, user_id: str) -> bool:
+        """Return whether user is setup."""
+        if self._users is None:
+            await self._async_load()
+
+        for user in self._users:  # type: ignore
+            if user[STORAGE_USER_ID] == user_id:
+                return True
+        return False
+
+    async def async_validation(
+            self, user_id: str, user_input: Dict[str, Any]) -> bool:
+        """Return True if validation passed."""
+        if self._users is None:
+            await self._async_load()
+
+        # user_input has been validate in caller
+        return await self.hass.async_add_executor_job(
+            self._validate_2fa, user_id, user_input[INPUT_FIELD_CODE])
+
+    def _validate_2fa(self, user_id: str, code: str) -> bool:
+        """Validate two factor authentication code."""
+        if user_id is None or code is None:
+            return False
+
+        import pyotp
+
+        ota_secret = None
+        for user in self._users:  # type: ignore
+            if user_id == user[STORAGE_USER_ID]:
+                ota_secret = user[STORAGE_OTA_SECRET]
+                break
+
+        if ota_secret is None:
+            # even we cannot find user, we still do verify
+            # to make timing the same as if user was found.
+            pyotp.TOTP(pyotp.random_base32()).verify(code)
```

Comment:
Is the time for generating the ota_secret via `random_base32` insignificant compared to the rest of the validation?

Reply:
Change to a fixed dummy code